sklearn_numba_dpex/kmeans/kernels.py

```diff
@@ -0,0 +1,807 @@
+                            window_feature_idx, window_centroid_idx
+                        ]
+                        partial_dot_products[window_centroid_idx] += (
+                            centroid_value * X_value
+                        )
+                    # The l2 norm of the current sample is needed to compute the exact
+                    # value of the inertia. It is accumulated in the first iteration
+                    # of the outer loop
+                    if _0 == 0:
+                        X_l2_norm += X_value * X_value
+
+                dpex.barrier(dpex.CLK_LOCAL_MEM_FENCE)
+
+                first_feature_idx += window_n_features
+
+            for i in range(window_n_centroids):
+                current_sample_pseudo_inertia = (
+                    centroids_window_half_l2_norm[i] - partial_dot_products[i]
+                )
+                if current_sample_pseudo_inertia < min_sample_pseudo_inertia:
+                    min_sample_pseudo_inertia = current_sample_pseudo_inertia
+                    min_idx = first_centroid_idx + i
+
+            dpex.barrier(dpex.CLK_LOCAL_MEM_FENCE)
+
+            first_centroid_idx += window_n_centroids
+
+        # No update step, only store min_idx in the output array
+        if sample_idx >= n_samples:
+            return
+
+        assignments_idx[sample_idx] = min_idx
+        per_sample_inertia[sample_idx] = X_l2_norm + (two * min_sample_pseudo_inertia)
+
+    return assignment[global_size, work_group_size]
+
+
+# TODO: all auxilliary kernels thereafter might be better optimized and we could
+# benchmark alternative implementations for each of them, that could include
+#    - using 2D or 3D grid of work groups and work items where applicable (e.g. in
+# some of the kernels that take 2D or 3D data as input) rather than using 1D grid. When
+# doing so, one should be especially careful about how the segments of adjacent work
+# items of size preferred_work_group_size_multiple are dispatched especially regarding
+# RW  operations in memory. A wrong dispatch strategy could slash memory bandwith and
+# reduce performance. Using 2D or 3D grid correctly might on the other hand improve
+# performance since it saves costly indexing operations (like //)
+#    - investigate if flat 1D-like indexing also works for ND kernels, thus saving the
+# need to compute coordinates for each dimension for element-wise operations.
+#    - or using numba + dpnp to directly leverage kernels that are shipped in dpnp to 
+# replace numpy methods.
+# However, in light of our main goal that is bringing a GPU KMeans to scikit-learn, the 
+# importance of those TODOs is currently seen as secondary, since the execution time of 
+# those kernels is only a small fraction of the total execution time and the
+# improvements that further optimizations can add will only be marginal. There is no
+# doubt, though, that a lot could be learnt about kernel programming in the process.
```

Comment:
:100:

I think we agree that subsequent optimisations should be made greedily based on profiling and observing the part of the implementation which takes more time to run and have a potential for optimisation.